// surrealdb.js/esm/library/SurrealSocket.js
import { WebsocketStatus, } from "../types.js";
import WebSocket from "./WebSocket/node.js";
import { getIncrementalID } from "./getIncrementalID.js";
import { processUrl } from "./processUrl.js";
export class SurrealSocket {
    constructor({ url, onOpen, onClose, }) {
        Object.defineProperty(this, "url", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "onOpen", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "onClose", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "ws", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "status", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: WebsocketStatus.CLOSED
        });
        Object.defineProperty(this, "queue", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: {}
        });
        Object.defineProperty(this, "ready", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "resolveReady", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "closed", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "resolveClosed", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "socketClosureReason", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: {
                1000: "CLOSE_NORMAL",
            }
        });
        this.resolveReady = () => { }; // Purely for typescript typing :)
        this.ready = new Promise((r) => this.resolveReady = r);
        this.resolveClosed = () => { }; // Purely for typescript typing :)
        this.closed = new Promise((r) => this.resolveClosed = r);
        this.onOpen = onOpen;
        this.onClose = onClose;
        this.url = processUrl(url, {
            http: "ws",
            https: "wss",
        }) + "/rpc";
    }
    open() {
        // Close any possibly connected sockets, reset status;
        this.close(1000);
        this.resetReady();
        // Connect to Surreal instance
        this.ws = new WebSocket(this.url);
        this.ws.addEventListener("open", (_e) => {
            this.status = WebsocketStatus.OPEN;
            this.resolveReady();
            this.onOpen?.();
        });
        this.ws.addEventListener("close", (_e) => {
            this.resolveClosed();
            this.resetClosed();
            // Connection retry mechanism
            if (this.status !== WebsocketStatus.CLOSED) {
                this.status = WebsocketStatus.RECONNECTING;
                setTimeout(() => {
                    this.open();
                }, 2500);
                this.onClose?.();
            }
        });
        this.ws.addEventListener("message", (e) => {
            const res = JSON.parse(e.data.toString());
            if (res.id && res.id in this.queue) {
                this.queue[res.id](res);
                delete this.queue[res.id];
            }
        });
    }
    // Extracting the pure object to prevent any getters/setters that could break stuff
    // Prevent user from overwriting ID that is being sent
    async send(method, params, callback) {
        await this.ready;
        const id = getIncrementalID();
        this.queue[id] = callback;
        this.ws?.send(JSON.stringify({ id, method, params }));
    }
    async close(reason) {
        this.status = WebsocketStatus.CLOSED;
        this.ws?.close(reason, this.socketClosureReason[reason]);
        this.onClose?.();
        await this.closed;
    }
    get connectionStatus() {
        return this.status;
    }
    resetReady() {
        this.ready = new Promise((r) => (this.resolveReady = r));
    }
    resetClosed() {
        this.closed = new Promise((r) => (this.resetClosed = r));
    }
}
